src/persist/src/indexed/encoding.rs

```diff
@@ -190,24 +195,48 @@ impl BlobMeta {
             }
             ids.insert(*id);
         }
-        for (_, f) in self.futures.iter() {
+        for (id, f) in self.futures.iter() {
+            if !ids.contains(id) {
+                return Err(format!("futures id {:?} not present in id_mapping", id).into());
+            }
             f.validate()?;
         }
-        for (_, t) in self.traces.iter() {
+        for (id, t) in self.traces.iter() {
+            if !ids.contains(id) {
+                return Err(format!("traces id {:?} not present in id_mapping", id).into());
+            }
             t.validate()?;
         }
-        // TODO: Assert that each stream id in futures is also present in traces
-        // and vice-versa. Also validate that that ts_lower of each future lines
-        // up wth the ts_upper of the corresponding trace.
+        let futures: HashMap<_, _> = self.futures.iter().map(|(id, m)| (*id, m)).collect();
+        let traces: HashMap<_, _> = self.traces.iter().map(|(id, m)| (*id, m)).collect();
+        for id in ids.iter() {
+            let future = futures.get(id).ok_or_else(|| {
+                Error::from(format!("id_mapping id {:?} not present in futures", id))
+            })?;
+            let trace = traces.get(id).ok_or_else(|| {
+                Error::from(format!("id_mapping id {:?} not present in traces", id))
+            })?;
+            let trace_ts_upper = trace.batches.last().map_or_else(
+                || Antichain::from_elem(Timestamp::minimum()),
+                |(d, _)| d.upper().clone(),
+            );
+            if trace_ts_upper != future.ts_lower {
+                return Err(Error::from(format!(
+                    "id {:?} trace ts_upper {:?} does not match future ts_lower {:?}",
+                    id, trace_ts_upper, future.ts_lower,
+                )));
+            }
+        }
         Ok(())
     }
+}
 
-    /// Returns the current SystemTime in millis since epoch.
-    pub fn now_millis() -> u128 {
-        SystemTime::now()
```

Comment:
Mental TODO I made to get rid of this when Matt did his SystemTime::now codebase purge. A counter is just as good and simpler